lib/auth/idp.go

```diff
@@ -0,0 +1,149 @@
+limitations under the License.
+*/
+
+package auth
+
+import (
+	"context"
+	"encoding/xml"
+	"net/url"
+
+	"github.com/beevik/etree"
+	"github.com/crewjam/saml"
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/auth/keystore"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/teleport/lib/tlsca"
+)
+
+// ProcessSAMLIdPRequest makes a signed SAML response to a SAML auth request.
+//
+//nolint:revive // Because we want this to be IdP.
+func (s *Server) ProcessSAMLIdPRequest(ctx context.Context, req *proto.ProcessSAMLIdPRequestRequest) (*proto.ProcessSAMLIdPRequestResponse, error) {
+	return ProcessSAMLIdPRequest(ctx, s, s.keyStore, req)
+}
+
+// ProcessSAMLIdPRequestClient is a client for retrieving CAs and cluster names, which is
+// necessary for creating SAML responses.
+type ProcessSAMLIdPRequestClient interface {
+	services.AuthorityGetter
+
+	// GetClusterName returns the local cluster name
+	GetClusterName(...services.MarshalOption) (types.ClusterName, error)
+
+	// ListSAMLIdPServiceProviders returns a paginated list of SAML IdP service provider resources.
+	ListSAMLIdPServiceProviders(ctx context.Context, pageSize int, nextToken string) ([]types.SAMLIdPServiceProvider, string, error)
+}
+
+// ProcessSAMLIdPRequest will make a signed SAML response to a SAML auth request.
+// This function is exported for testing in the IdP itself.
+//
+//nolint:revive // Because we want this to be IdP.
+func ProcessSAMLIdPRequest(ctx context.Context, client ProcessSAMLIdPRequestClient, keyStore *keystore.Manager, req *proto.ProcessSAMLIdPRequestRequest) (*proto.ProcessSAMLIdPRequestResponse, error) {
+	// Get the cert for signing the request.
+	clusterName, err := client.GetClusterName()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	ca, err := client.GetCertAuthority(ctx, types.CertAuthID{
+		Type:       types.SAMLIDPCA,
+		DomainName: clusterName.GetClusterName(),
+	}, true)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	rawCert, signer, err := keyStore.GetTLSCertAndSigner(ctx, ca)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	cert, err := tlsca.ParseCertificatePEM(rawCert)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Parse the assertion XML
+	assertionDoc := etree.NewDocument()
+	if err := assertionDoc.ReadFromBytes(req.Assertion); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	assertion := &saml.Assertion{}
+	if err := xml.Unmarshal(req.Assertion, assertion); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Get the metadata URL, which for our IdP implementation doubles as the
+	// entity ID.
+	metadataURL, err := url.Parse(req.MetadataURL)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Create an IdP object to stuff into the IdpAuthnRequest.
+	idp := &saml.IdentityProvider{
+		Certificate:     cert,
+		Key:             signer,
```

Comment:
as we discussed I think this still isn't going to work with HSM keys due to lack of support in the library, but at least having it here (on Auth) gives us the option to make that work and keeps all CA signing operations on Auth